src/BitcoinLightClient.sol

```diff
@@ -0,0 +1,273 @@
+        bytes32 prevBlock,
+        bytes32 merkleRoot
+    ) public initializer {
+        __AccessControl_init();
+        __UUPSUpgradeable_init();
+
+        // Setup roles
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+        _grantRole(UPGRADER_ROLE, admin);
+
+        // Initialize first checkpoint
+        BitcoinUtils.BlockHeader memory header = BitcoinUtils.BlockHeader({
+            version: blockVersion,
+            timestamp: blockTimestamp,
+            difficultyBits: difficultyBits,
+            nonce: nonce,
+            height: height,
+            prevBlock: prevBlock,
+            merkleRoot: merkleRoot
+        });
+
+        latestCheckpointHeaderHash = BitcoinUtils.getBlockHashFromParams(header);
+        headers[latestCheckpointHeaderHash] = header;
+    }
+
+    /**
+     * @notice Submit a new block header fields along with intermediate headers
+     * @param blockVersion Block version
+     * @param blockTimestamp Block timestamp
+     * @param difficultyBits Block difficulty bits
+     * @param nonce Block nonce
+     * @param height Block height
+     * @param prevBlock Previous block hash
+     * @param merkleRoot Block merkle root
+     * @param intermediateHeaders Array of intermediate headers
+     * @dev Only accounts with BLOCK_SUBMIT_ROLE can submit headers
+     */
+    function submitBlockHeader(
+        uint32 blockVersion,
+        uint32 blockTimestamp,
+        uint32 difficultyBits,
+        uint32 nonce,
+        uint32 height,
+        bytes32 prevBlock,
+        bytes32 merkleRoot,
+        bytes[] calldata intermediateHeaders
+    ) external returns (bool) {
+        BitcoinUtils.BlockHeader memory header = BitcoinUtils.BlockHeader({
+            version: blockVersion,
+            timestamp: blockTimestamp,
+            difficultyBits: difficultyBits,
+            nonce: nonce,
+            height: height,
+            prevBlock: prevBlock,
+            merkleRoot: merkleRoot
+        });
+        bytes32 blockHash = BitcoinUtils.getBlockHashFromParams(header);
+        return _submitBlockHeader(blockHash, header, intermediateHeaders);
+    }
+
+    /**
+     * @notice Submit a new raw block header along with intermediate headers
+     * @param rawHeader Raw block header bytes
+     * @param intermediateHeaders Array of intermediate headers
+     * @dev Only accounts with BLOCK_SUBMIT_ROLE can submit headers
+     */
+    function submitRawBlockHeader(bytes calldata rawHeader, bytes[] calldata intermediateHeaders)
+        external
+        returns (bool)
+    {
+        bytes32 blockHash = getBlockHash(rawHeader);
+        BitcoinUtils.BlockHeader memory header = BitcoinUtils.parseBlockHeader(rawHeader);
+        return _submitBlockHeader(blockHash, header, intermediateHeaders);
+    }
+
+    /**
+     * @notice Internal function to submit block header
+     * @param blockHash Block hash
+     * @param header Block header
+     * @param intermediateHeaders Array of intermediate headers
```

Comment:
please add the assumption that intermediate Headers are in the reverse order. Also add the example reference of it to understand the functionality of the function